### PR TITLE
[DRAFT] Refactor models, create extensible AWQ framework

### LINKED ISSUE #71: Create models directory in AWQ with model classes
Ideally, when trying to support more models, you should create classes for each model to better manage them. This is an open issue to describe what should be moved into model classes in order to abstract the code a bit further and make it easier to support new models.

- Extract the [get_blocks](https://github.com/mit-han-lab/llm-awq/blob/main/awq/quantize/pre_quant.py#L22) function into model class
- Extract the [move_embed](https://github.com/mit-han-lab/llm-awq/blob/main/awq/quantize/pre_quant.py#L37) function into model class
- Extract the [scale_activations](https://github.com/mit-han-lab/llm-awq/blob/main/awq/quantize/quantizer.py#L14) function into model class
- Extract part of [auto_scale_block](https://github.com/mit-han-lab/llm-awq/blob/main/awq/quantize/auto_scale.py#L178) function into model class

### PR DESCRIPTION
**Pull request by casper-hansen**
Work in progress—attempts to resolve #71.

- [x] Create base class
    - [x] Remove `pre_quant.py`
- [X] Create auto class
    - [x] from_quantized()
    - [x] from_pretrained()
- [x] Create model classes
    - [x] MPT
    - [x] LLaMa
    - [x] OPT
    - [x] Bloom
    - [x] Falcon
- [x] Implement base methods
    - [x] quantize()
    - [x] save_quantized()
    - [x] from_pretrained()
    - [x] from_quantized()
- [x] Refactor `entry.py`
- [x] Adapt TinyChat to use Auto classes
- [x] Create and load `quant_config.json`

### CODE REVIEW

**Comment by abhinavkulkarni** [comment]
@casperbh96: Why move around so much code?

All we need is a module that has all the `Auto` method implemented. As for loading the model, you can take a look at any of the models I have released [here](https://huggingface.co/abhinavkulkarni?sort_models=downloads#models). We just need to:

1. read `quantize_config.json`
2. initialize an empty model (on `meta` device) using model config
3. call `real_quantize_model_weight` on the model
4. finally load the wait from the directory passed

Everything else can remain as is.

Let me know what you think.

**Comment by casperbh96** [comment]
Let me start by saying this. The motivation for this entire PR is to make AWQ easier to work with. To do this, I adopted a design similar to AutoGPTQ, which has been touted as having a great design and has also been integrated into huggingface/transformers. Although AutoGPTQ has a great framework to work within, the GPTQ method is unfortunately slow compared to AWQ. So I think of it as AutoAWQ but with superior speed.

> Why move around so much code?

It is mainly the code from `pre_quant.py` and `quantizer.py` that has been moved. I also resolved some circular imports that occurred when moving code - mostly due to how it was structured before.

1. `scale_activations` needs to be in `base.py` because it relies on the model classes implemented.
2. `pseudo_quantize_model_weight` was unused and `real_quantize_model_weight` was refactored into the `quantize` and `from_quantized` method.
3. `auto_scale` code for models is moved into model classes
4. `pre_quant.py` had a lot of code pertaining to what should happen in the case of a specific model, so also moved into model classes while `run_awq` is the search method of AWQ that was moved into `quantize`.

The idea of this PR was to make it easy to support more models, and in order to do that, I needed to create methods in base, auto, mpt classes to make sure it's easy to work with. This is similar to how AutoGPTQ is implemented, and it helps a lot when understanding how the quantization code works. Before, it was scattered around, but now the init is from the `quantize` method in the base model.

In summary, a lot of model code and quantization code were scattered around in AWQ, and now it has been neatly organized.

> from_quantized logic

All steps that you mentioned are implemented except for the `quantize_config.json` and loading to `meta` device. I will be doing more cleanup and providing as many comments as I can on the methods in `base.py` so that it is 1) easy to understand, 2) easy to maintain, and 3) easy to build upon.

**Comment by casperbh96** [comment]
@tonylins @Sakits @kentang-mit 

Please review this PR.

I have successfully reworked the framework now and modified the examples. Everything works as before, it is just a little easier now.

The main idea in this pull request is to make an AWQ framework that is easy to extend and work with. This is seen in how you can easily quantize a model and load a model afterwards. Please inspect the new `entry.py` and see the simplicity in how it works.

For example, running inference will be very easy now, just a couple of lines. 

```python
from transformers import AutoTokenizer
from awq.models.auto import AutoAWQForCausalLM

quant_path = "casperhansen/vicuna-7b-v1.5-awq"
quant_file = "awq_model_w4_g128.pt"

model = AutoAWQForCausalLM.from_quantized(quant_path, quant_file)
tokenizer = AutoTokenizer.from_pretrained(quant_path, trust_remote_code=True)

model.generate(...)
```

**Comment by tonylins** [comment]
Hi @casperbh96 , thanks for the big PR! @Sakits @kentang-mit and I will review the PR and do some discussions/testings in these two days.

**Comment by kentang-mit** [comment]
Thanks @casperbh96 for your efforts! We will have a look at the PR~

**Comment by casper-hansen** [comment]
Hi @tonylins @kentang-mit @Sakits, I have decided to close this pull request so that you can preserve your original work.

I have instead created another repository to work on further optimizing AWQ and making it easier to use. 
https://github.com/casper-hansen/AutoAWQ

**Comment by kentang-mit** [comment]
Thank you for the efforts, @casper-hansen! I am still in the process of learning your great work, but really appreciate your contribution!

**Comment by casper-hansen** [comment]
> Thank you for the efforts, @casper-hansen! I am still in the process of learning your great work, but really appreciate your contribution!

Thank you for creating the research around AWQ. The goal of AutoAWQ is to get integrated into the Huggingface ecosystem and expand support such that AWQ can be adopted more widely. One of the things that are high on my priority list is optimizing the GEMM CUDA kernel for more performance, for instance, I found #80 pretty exciting.

Another exciting thing is to support BERT models. It's a bit complicated, but I believe it can be done if the code is further refactored. The problem is that BERT models do not have a `nn.Linear` or `nn.LayerNorm` to reference for the `attn_in` like in other models, so I am looking into how this can be solved to add support for it. 

You can see the initial code I wrote for BERT below:
https://github.com/casper-hansen/AutoAWQ/blob/bert/awq/models/bert.py#L37

**Comment by tonylins** [comment]
Hi @casper-hansen Thanks for the effort! Do you want to add a reference in our README so that people can know about the work? Feel free to create a PR when you feel ready.

**Comment by casper-hansen** [comment]
@tonylins Thank you for your support. Let me do that, thanks :)
